### PR TITLE
Merged blocks distribution

### PR DESCRIPTION
**Pull request by eranrund**
### Motivation

We'd like to reduce the amount of requests necessary to obtain blocks data in order to speed up catchup on slower connections.

### In this PR
* Introduce the concept of "merged blocks" - these are objects that hold an array of consecutive blocks.
* Change `mc-ledger-distribution` to support uploading of merged blocks every 100/1000/10000 blocks. (e.g. when block 99 is completed, blocks 0-99 would be joined into a merged block and uploaded to the storage provider, in addition to having all the individual blocks. When block 9999 is completed, blocks 0-9999 (10k merged blocks), 9000-9999 (1k merged blocks) and 9900-9999 (100 merged blocks) would be stored in addition to the individual blocks.
* Change `RequestTransactionsFetcher` to support getting the merged blocks and using them as cache.

Some notes:
* This has no effect on ledger format, the blockchain, or anything other than the distribution mechanism.
* In order to benefit from this in the existing test net, `mc-ledger-distribution` would need to be reran, starting from block zero. This will force it to create the merged blocks objects. For a new network that is started from block 0, this should work seamlessly
* When testing performance locally, this performs similarly to the current implementation but with less concurrency needed. I expect this will pay higher dividends on slower connections, where the overhead negotiating a TLS connection for each block is noticeable.

### CODE REVIEW

File: api/proto/blockchain.proto [hunk 1]
```diff
@@ -80,3 +80,8 @@ message ArchiveBlock {
         ArchiveBlockV1 v1 = 1;
     }
 }
+
+// A collection of archived blocks.
+message ArchiveBlocks {
```
Comment (eranrund) [inline api/proto/blockchain.proto:85]: aka what I refer to as "merged blocks" in other places.

File: api/src/convert/archive_block.rs [hunk 2]
```diff
@@ -63,3 +65,50 @@ impl TryFrom<&blockchain::ArchiveBlock> for mc_transaction_core::BlockData {
         ))
     }
 }
+
+/// Convert &[mc_transaction_core::BlockData] -> blockchain::ArchiveBlocks
+impl From<&[mc_transaction_core::BlockData]> for blockchain::ArchiveBlocks {
```
Comment (eranrund) [inline api/src/convert/archive_block.rs:70]: I still need to write tests for these.

**Review by sugargoat — APPROVED** [review]
Very excited about this change and the follow-ons! LGTM

File: api/src/convert/archive_block.rs [hunk 3]
```diff
@@ -63,3 +65,207 @@ impl TryFrom<&blockchain::ArchiveBlock> for mc_transaction_core::BlockData {
...
+    // Attempting to convert an ArchiveBlock with invalid signature or contents should fail.
+    fn try_from_blockchain_archive_block_rejects_invalid() {
+        let block_data = generate_test_blocks_data(1).pop().unwrap();
+
+        // ArchiveBlock with invalid signature cannot be converted back to BlockData
+        let mut archive_block = blockchain::ArchiveBlock::from(&block_data);
+        archive_block
+            .mut_v1()
+            .mut_signature()
+            .mut_signature()
```
Comment (rjwalters) [inline api/src/convert/archive_block.rs:210]: 2x mut_signature?

File: ledger/sync/src/reqwest_transactions_fetcher.rs [hunk 4]
```diff
@@ -53,10 +66,31 @@ impl TransactionFetcherError for ReqwestTransactionsFetcherError {}
...
+    /// List of URLs to try and fetch objects from.
     pub source_urls: Vec<Url>,
+
+    /// Client used for HTTP(s) requests.
     client: reqwest::blocking::Client,
+
+    /// Logger.
     logger: Logger,
+
+    /// The most recently used URL  index (in `source_urls`).
```
Comment (rjwalters) [inline ledger/sync/src/reqwest_transactions_fetcher.rs:78]:
```suggestion
    /// The most recently used URL index (in `source_urls`).
```

**Review by rjwalters — APPROVED** [review]

File: api/src/convert/archive_block.rs (hunk 3 above)
Comment (eranrund) [reply api/src/convert/archive_block.rs:210]: Yes. 
https://github.com/mobilecoinofficial/mobilecoin/blob/master/api/proto/blockchain.proto#L74
https://github.com/mobilecoinofficial/mobilecoin/blob/master/api/proto/blockchain.proto#L55

**Pull request merged**